core/index/heap_iterator.hpp

```diff
@@ -30,77 +30,96 @@
 
 namespace irs {
 
-// External heap iterator
-// ----------------------------------------------------------------------------
-//      [0] <-- begin
-//      [1]      |
-//      [2]      | head (heap)
-//      [3]      |
-//      [4] <-- lead_
-//      [5]      |
-//      [6]      | lead (list of accepted iterators)
-//      ...      |
-//      [n] <-- end
-// ----------------------------------------------------------------------------
 template<typename Context>
-class ExternalHeapIterator {
+class ExternalMergeIterator {
+  using Value = typename Context::Value;
+
  public:
-  explicit ExternalHeapIterator(Context&& ctx) : ctx_{std::move(ctx)} {}
+  template<typename... Args>
+  explicit ExternalMergeIterator(Args&&... args)
+    : ctx_{std::forward<Args>(args)...} {}
+
+  bool Initilized() const noexcept { return !tree_.empty(); }
 
-  void reset(size_t size) {
-    heap_.resize(size);
-    std::iota(heap_.begin(), heap_.end(), size_t{0});
-    lead_ = size;
+  void Reset(std::span<Value> values) noexcept {
+    size_ = 0;
+    tree_.clear();
+    values_ = values;
   }
 
-  bool next() {
-    if (heap_.empty()) {
-      return false;
+  bool Next() {
+    if (IRS_UNLIKELY(size_ == 0)) {
+      return LazyReset();
     }
-
-    auto begin = std::begin(heap_);
-
-    while (lead_) {
-      auto it = std::end(heap_) - lead_--;
-
-      if (!ctx_(*it)) {  // advance iterator
-        if (!remove_lead(it)) {
-          IRS_ASSERT(heap_.empty());
-          return false;
-        }
-
-        continue;
+    auto& lead = Lead();
+    auto position = values_.size() + (&lead - values_.data());
+    if (IRS_UNLIKELY(!ctx_(lead))) {
+      if (--size_ == 0) {
+        return false;
       }
-
-      std::push_heap(begin, ++it, ctx_);
+      tree_[position] = nullptr;
+    }
+    while ((position /= 2) != 0) {
+      tree_[position] = Compute(position);
     }
-
-    IRS_ASSERT(!heap_.empty());
-    std::pop_heap(begin, std::end(heap_), ctx_);
-    lead_ = 1;
-
     return true;
   }
 
-  size_t value() const noexcept {
-    IRS_ASSERT(!heap_.empty());
-    return heap_.back();
+  IRS_FORCE_INLINE Value& Lead() const noexcept {
+    IRS_ASSERT(size_ > 0);
+    IRS_ASSERT(tree_.size() > 1);
+    IRS_ASSERT(tree_[1] != nullptr);
+    return *tree_[1];
   }
 
-  size_t size() const noexcept { return heap_.size(); }
+  size_t Size() const noexcept { return size_; }
 
  private:
-  bool remove_lead(std::vector<size_t>::iterator it) {
-    if (&*it != &heap_.back()) {
-      std::swap(*it, heap_.back());
+  IRS_FORCE_INLINE Value* Compute(size_t position) {
+    auto* lhs = tree_[2 * position];
+    auto* rhs = tree_[2 * position + 1];
+    if ((lhs != nullptr) & (rhs != nullptr)) {
+      return ctx_(*lhs, *rhs) ? lhs : rhs;
+    }
+    return reinterpret_cast<Value*>(reinterpret_cast<uintptr_t>(lhs) |
+                                    reinterpret_cast<uintptr_t>(rhs));
+  }
+
+  IRS_NO_INLINE bool LazyReset() {
```

Comment:
Why force no-inline here?

Reply:
Because it's called single time, only on first Next.
So better to doesn't have it code in Next or in cycle with Next
Also it will be tail call, so it just jmp